File: leetcode_top_interview/src/leetcode_top_interview/maximal_square.rs
struct Solution;
impl Solution {
    #[allow(dead_code)]
    pub fn maximal_square(matrix: Vec<Vec<char>>) -> i32 {
        let mut matrix = matrix
            .into_iter()
            .map(|x| {
                x.into_iter()
                    .map(|x| x as i32 - '0' as i32)
                    .collect::<Vec<_>>()
            })
            .collect::<Vec<_>>();
        let row_num = matrix.len();
        let col_num = matrix[0].len();
        for row in 1..row_num {
            for col in 1..col_num {
                if matrix[row][col] == 1 {
                    let max_square = matrix[row][col - 1]
                        .min(matrix[row - 1][col])
                        .min(matrix[row - 1][col - 1])
                        + 1;
                    matrix[row][col] = max_square;
                }
            }
        }
        matrix
            .iter()
            .map(|x| x.iter().max().map(|y| *y * *y).unwrap())
            .max()
            .unwrap()
    }
}
#[cfg(test)]
mod test_super {
    use super::*;

    #[test]
    fn test_1() {
        let matrix = [
            ["1", "0", "1", "0", "0"],
            ["1", "0", "1", "1", "1"],
            ["1", "1", "1", "1", "1"],
            ["1", "0", "0", "1", "0"],
        ]
        .iter()
        .map(|x| {
            x.iter()
                .map(|y| y.chars().next().unwrap())
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();
        assert_eq!(4, Solution::maximal_square(matrix));
    }
}

File: leetcode_cpp/Dynamic Programming/Matrix/maximalSquare.cpp
#include <bits/stdc++.h>
using namespace std;
auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();

class Solution {
  public:
    static int maximalSquare(vector<vector<char>>& matrix) {
        int ret = 0;
        int row = matrix.size();
        int col = matrix[0].size();
        std::vector<std::vector<int>> imatrix(row, std::vector<int>(col, 0));
        for (int i = 0; i < row; ++i) {
            for (int j = 0; j < col; ++j) {
                imatrix[i][j] = matrix[i][j] - '0';
                ret = std::max(imatrix[i][j], ret);
            }
        }
        for (int i = 1; i < row; ++i) {
            for (int j = 1; j < col; ++j) {
                if (imatrix[i][j] == 1) {
                    imatrix[i][j] = std::min({imatrix[i - 1][j], imatrix[i][j - 1], imatrix[i - 1][j - 1]}) + 1;
                    ret = std::max(ret, imatrix[i][j]);
                }
            }
        }
        return ret * ret;
    }
};